components/builder-worker/src/runner/mod.rs

```diff
@@ -0,0 +1,337 @@
+use std::str::FromStr;
+use std::sync::{mpsc, Arc, RwLock};
+use std::thread::{self, JoinHandle};
+
+use depot_client;
+use hab_core::package::archive::PackageArchive;
+use hab_core::package::install::PackageInstall;
+use hab_core::package::PackageIdent;
+use hab_net::server::ZMQ_CONTEXT;
+use protobuf::{parse_from_bytes, Message};
+use protocol::jobsrv as proto;
+use zmq;
+
+use self::logger::Logger;
+use self::workspace::Workspace;
+use config::Config;
+use error::{Error, Result};
+use vcs;
+use {PRODUCT, VERSION};
+
+/// In-memory zmq address of Job RunnerMgr
+const INPROC_ADDR: &'static str = "inproc://runner";
+/// Protocol message to indicate the Job Runner has received a work request
+const WORK_ACK: &'static str = "A";
+/// Protocol message to indicate the Job Runner has completed a work request
+const WORK_COMPLETE: &'static str = "C";
+
+lazy_static! {
+    static ref STUDIO_PKG: PackageIdent = PackageIdent::from_str("core/hab-studio").unwrap();
+}
+
+pub struct Runner {
+    pub workspace: Workspace,
+    auth_token: String,
+}
+
+impl Runner {
+    pub fn new(job: proto::Job, config: &Config) -> Self {
+        Runner {
+            auth_token: config.auth_token.clone(),
+            workspace: Workspace::new(config.data_path.clone(), job),
+        }
+    }
+
+    pub fn run(&mut self) -> () {
+        if let Some(err) = self.setup().err() {
+            error!("WORKSPACE SETUP ERR={:?}", err);
+            return self.fail();
+        }
+        let mut logger = Logger::new(&self.workspace);
+        // JW TODO: How are we going to get the secret keys for this thing?
+        if let Some(err) = vcs::clone(&self.workspace.job, &self.workspace.src()).err() {
+            error!("CLONE ERROR={}", err);
+            return self.fail();
+        }
+        let mut archive = match self.build(&mut logger) {
+            Ok(archive) => archive,
+            Err(err) => {
+                error!("STUDIO ERR={}", err);
+                return self.fail();
+            }
+        };
+        if !self.post_process(&mut archive) {
+            // JW TODO: We should shelve the built artifacts and allow a retry on post-processing.
+            // If the job is killed then we can kill the shelved artifacts.
+            return self.fail();
+        }
+        self.complete()
+    }
+
+    pub fn build(&self, logger: &mut Logger) -> Result<PackageArchive> {
+        let args =
+            vec![OsString::from("-s"),
+                 OsString::from(self.workspace.src()),
+                 OsString::from("-r"),
+                 OsString::from(self.workspace.studio()),
+                 OsString::from("-k"),
+                 OsString::from("core"),
+                 OsString::from("build"),
+                 OsString::from(Path::new(self.workspace.job.get_project().get_plan_path())
+                     .parent()
+                     .unwrap())];
+        let command = match PackageInstall::load(&STUDIO_PKG, None) {
```

Comment:
Yeah I looked at this one for a bit before deciding to defer to later. We need to expose an API in core which allows our Rust code to identify if it, itself, was packaged with Habitat and then we need an API to retrieve the exact dep for a given `origin`/`name` combination. I didn't want to tackle that work just yet, so I left it with this which should be good enough for our acceptance environment.

I just pushed some changes which document this to be fixed